src/cli/expand-patterns.js

```diff
@@ -0,0 +1,221 @@
+   *  path?: string;
+   *  glob?: string;
+   *  input: string;
+   * }>}
+   */
+  const entries = [];
+
+  for (const pattern of context.filePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+
+    if (containsIgnoredPathSegment(absolutePath, cwd, ignoredDirectories)) {
+      continue;
+    }
+
+    const stat = statSafeSync(absolutePath);
+    if (stat) {
+      if (stat.isFile()) {
+        entries.push({ type: "file", path: absolutePath, input: pattern });
+      } else if (stat.isDirectory()) {
+        entries.push({
+          type: "dir",
+          path: absolutePath,
+          glob: getSupportedFilesGlob(),
+          input: pattern
+        });
+      }
+    } else if (pattern[0] === "!") {
+      // convert negative patterns to `ignore` entries
+      globbyOptions.ignore.push(pattern.slice(1));
+    } else {
+      entries.push({
+        type: "glob",
+        // Using backslashes in globs is probably not okay, but not accepting
+        // backslashes as path separators on Windows is even more not okay.
+        // https://github.com/prettier/prettier/pull/6776#discussion_r380723717
+        // https://github.com/mrmlnc/fast-glob#how-to-write-patterns-on-windows
+        glob: isWindows ? pattern.replace(/\\/g, "/") : pattern,
+        input: pattern
+      });
+    }
+  }
+
+  for (const { type, path, glob, input } of entries) {
+    switch (type) {
+      case "file":
+        yield path;
+        continue;
+
+      case "dir": {
+        let result;
+        try {
+          result = globby.sync(glob, { ...globbyOptions, cwd: path });
+        } catch ({ message }) {
+          yield { error: `Unable to expand directory: ${input}\n${message}` };
+          continue;
+        }
+
+        if (result.length === 0) {
+          yield {
+            error: `No supported files were found in the directory "${input}".`
+          };
+        } else {
+          yield* sortIfNotBig(result);
+        }
+        continue;
+      }
+
+      case "glob": {
+        let result;
+        try {
+          result = globby.sync(glob, globbyOptions);
+        } catch ({ message }) {
+          yield {
+            error: `Unable to expand glob pattern: ${input}\n${message}`
+          };
+          continue;
+        }
+
+        if (result.length === 0) {
+          yield {
+            error: `No files matching the pattern "${input}" were found.`
+          };
+        } else {
+          yield* sortIfNotBig(result);
+        }
+        continue;
+      }
+    }
+  }
+
+  function getSupportedFilesGlob() {
+    if (!supportedFilesGlob) {
+      const extensions = flat(
+        context.languages.map(lang => lang.extensions || [])
+      );
+      const filenames = flat(
+        context.languages.map(lang => lang.filenames || [])
+      );
+      supportedFilesGlob = `**/{${extensions
+        .map(ext => "*" + (ext[0] === "." ? ext : "." + ext))
+        .concat(filenames)}}`;
+    }
+    return supportedFilesGlob;
```

Comment:
Can I format  `foo.barjs` file with `config override `. I think right way is  glob all files
 and  `inferParser` then filter the files without  parser

Reply:
> Can I format exts add by config override or plugins?

Plugins - yes. There is a test for that.

Config overrides - no. Because configs are resolved on a later stage.

> I think we should glob all files and inferParser then filter the files without parser

That's a good idea. I'll give it a try.

Reply:
This parser inference thing looks a bit complicated though. How about merging this PR as is and improving this later?

I feel this PR is already good enough for most of the use cases. People just want to run `prettier . --write` or `prettier src --write`. For more complicated scenarios, they can use globs.